src/jupyterlabcontroller/services/prepuller/docker.py

```diff
@@ -1,34 +1,91 @@
-"""This class will manage communication with the Docker repository in order
-to fetch tag information and determine which images are available from the
-repository.
+"""Docker client for image prepuller."""
 
-These are known as our remote images.
-"""
+import asyncio
 
+from structlog.stdlib import BoundLogger
 
+from ...models.tag import RSPTag, RSPTagList, TagMap
+from ...models.v1.prepuller_config import PrepullerConfiguration
 from ...storage.docker import DockerStorageClient
 from .state import PrepullerState
 
 
 class PrepullerDockerClient:
+    """Query Docker for available images.
+
+    Fetches tag information for a given repository from a Docker registry.
+    These are known as our remote images and are used to calculate which
+    images to prepull.
+    """
+
     def __init__(
         self,
         namespace: str,
         state: PrepullerState,
         docker_client: DockerStorageClient,
+        config: PrepullerConfiguration,
+        logger: BoundLogger,
     ) -> None:
-        self.logger = docker_client.logger
+        self.logger = logger
         self.state = state
         self.namespace = namespace
         self.docker_client = docker_client
+        self.config = config
+
+    async def get_tag_map(self) -> TagMap:
+        """Get a list of all available tags.
+
+        Query all the tags for a given repository to get their digests,
+        inflate the tag and digest pairs into RSPTag objects, and return a
+        structure with dicts indexed by tag and by digest.  This in turn will
+        let us easily compare what we have remotely and what we have locally,
+        because it doesn't matter what tag we pulled a given image by, but
+        whether we have an image with the proper digest already on a given
+        node.
+        """
+        # Get all the tags and digests from the Docker API.  This is much too
+        # aggressive in parallelism and also is doing too much work since we
+        # don't care about the digests for most tags.
```

Comment:
I'm not at all sure we don't.  We will need it to determine which tags are aliases for other tags, and since we have it we can detect whether we have stale images stored locally.

Reply:
I've been having a hard time wrapping my mind around this. I think we care about the digests for everything we're prepulling in order to detect and replace stale images as you mention, and we care about mapping an alias to another tag for recommended and any other alias we want to give a special position in the menu. But I don't think we care about knowing which tags are aliases other than that, do we? The only place they're used is in the dropdown menu, which is just a list of every tag, and I don't think we care what is an alias and what isn't. We just list all the tags anyway.

This is probably a good stand-up topic, since there may be some subtlety here that I'm missing. But it feels weird to do hundreds of GET requests against the Docker API pretty frequently when we by and large don't use any of that data, only the digests for the much smaller set of prepulled images.